core/src/main/java/org/apache/druid/timeline/DataSegment.java

```diff
@@ -87,9 +100,9 @@
   private final Integer binaryVersion;
   private final SegmentId id;
   @Nullable
-  private final Map<String, Object> loadSpec;
-  private final List<String> dimensions;
-  private final List<String> metrics;
+  private volatile Map<String, Object> loadSpec;
```

Comment:
Making `DataSegment` non-immutable was ruled out in this discussion: #7571. Please read it in full. I proposed a solution here: https://github.com/apache/incubator-druid/issues/7571#issuecomment-495288920. Please check if you can implement it in this PR.